src/py21cmfast/wrapper/cfuncs.py

```diff
@@ -366,931 +434,393 @@ def construct_fftw_wisdoms(
+    growth = get_growth_factor(inputs=inputs, redshift=redshift)
+    return get_delta_crit_nu(inputs.user_params, sigma, growth)
+
+
+def get_delta_crit_nu(user_params, sigma, growth):
+    """Uses the nu paramters (sigma and growth factor) to get critical density."""
+    # None of the parameter structs are used in this function so we don't need a broadcast
+    return lib.get_delta_crit(user_params.cdict["HMF"], sigma, growth)
 
 
-def get_cmf_integral(
-    user_params: UserParams,
-    cosmo_params: CosmoParams,
-    astro_params: AstroParams,
-    flag_options: FlagOptions,
-    M_min: Sequence[float] | float,
-    M_max: Sequence[float] | float,
-    M_cond: Sequence[float] | float,
+def evaluate_condition_integrals(
+    inputs: InputParameters,
+    cond_array: Sequence[float],
     redshift: float,
-    delta: Sequence[float] | float | None = None,
-    z_desc: float | None = None,
+    redshift_prev: float | None = None,
 ):
-    """
-    Evaluates the simple halo mass function intgral.
+    """Gets the expected number and mass of halos given a condition.
 
-    Can be computed over a range of conditions and mass bounds.
+    If USE_INTERPOLATION_TABLES is set to 'hmf-interpolation': Will crash if the table
+    has not been initialised, only `cond_array` is used,
+    and the rest of the arguments are taken from when the table was initialised.
     """
-    lib.Broadcast_struct_global_all(
-        user_params.cstruct,
-        cosmo_params.cstruct,
-        astro_params.cstruct,
-        flag_options.cstruct,
+    cond_array = cond_array.astype("f8")
+    n_halo = np.zeros_like(cond_array)
+    m_coll = np.zeros_like(cond_array)
+
+    lib.get_condition_integrals(
+        inputs.user_params.cstruct,
+        inputs.cosmo_params.cstruct,
+        inputs.astro_params.cstruct,
+        inputs.flag_options.cstruct,
+        redshift,
+        redshift_prev if redshift_prev is not None else -1,
+        cond_array.size,
+        ffi.cast("double *", ffi.from_buffer(cond_array)),
+        ffi.cast("double *", ffi.from_buffer(n_halo)),
+        ffi.cast("double *", ffi.from_buffer(m_coll)),
     )
-    lib.init_ps()
-    lib.initialiseSigmaMInterpTable(1e5, 1e16)
 
-    sigma = np.vectorize(lib.EvaluateSigma)(np.log(M_cond))
+    return n_halo, m_coll
 
-    growth_out = lib.dicke(redshift)
-    if z_desc is not None and delta is None:
-        growth_in = lib.dicke(z_desc)
-        delta = (
-            (
-                np.vectorize(lib.get_delta_crit)(
-                    user_params.cdict["HMF"], sigma, growth_in
-                )
-            )
-            * growth_out
-            / growth_in
-        )
 
-    cmf_integral = np.vectorize(lib.Nhalo_Conditional)(
-        growth_out,
-        np.log(M_min),
-        np.log(M_max),
-        M_cond,
-        sigma,
-        delta,
-        0,  # GSL-QAG
+def integrate_chmf_interval(
+    inputs: InputParameters,
+    redshift: float,
+    lnM_lower: Sequence[float],
```

Comment:
I would prefer to maintain python conventions wherever possible, so use lower-case names `lnm_lower` etc. Also, should there be a check inside the function like `assert np.all(lnM_upper > lnM_lower)`?